osu.Game/Graphics/SpriteIcon.cs

```diff
@@ -60,10 +62,32 @@ private void updateTexture()
                 Size = new Vector2(texture?.DisplayWidth ?? 0, texture?.DisplayHeight ?? 0);
         }
 
+        public override bool Invalidate(Invalidation invalidation = Invalidation.All, Drawable source = null, bool shallPropagate = true)
+        {
+            if ((invalidation & Invalidation.Colour) > 0)
+            {
+                //adjust shadow alpha based on highest component intensity to avoid muddy display of darker text.
+                //squared result for quadratic fall-off seems to give the best result.
+                var avgColour = (Color4)DrawInfo.Colour.AverageColour;
+
+                spriteShadow.Alpha = (float)Math.Pow(Math.Max(Math.Max(avgColour.R, avgColour.G), avgColour.B), 2);
+            }
+
+            return base.Invalidate(invalidation, source, shallPropagate);
+        }
+
         public bool Shadow
         {
             get { return spriteShadow.IsPresent; }
-            set { spriteShadow.Alpha = value ? 1 : 0; }
+            set
+            {
+                if (value == (spriteShadow.IsAlive && spriteShadow.IsLoaded)) return;

```

Comment:
This is not non-async-load-safe. I can break this with the following code in a testcase ctor (or really, any ctor):
```
SpriteIcon sr;
Add(sr = new SpriteIcon
{
    Icon = FontAwesome.fa_question,
    Size = new Vector2(100)
});

for (int i = 0; i < 1000; i++)
    sr.Shadow = true;
```